### PR TITLE
feat: oracle log future processing

### PR DESCRIPTION
**Pull request by sokel**
This pr extends oracle functionality with new option:
now oracle would work with later transactions. This option is checked by default.

Reason: oracle in slave mode after restart process all logs in blockchain.

### CODE REVIEW

File: insonmnia/oracle/oracle.go [hunk 1]
```diff
@@ -85,8 +85,18 @@ func (o *Oracle) listenEventsRoutine(ctx context.Context) error {
 	if o.cfg.Oracle.IsMaster {
 		return nil
 	}
+
+	var lastBlock uint64 = 0
+	if o.cfg.Oracle.FromNow {
+		var err error
+		lastBlock, err = o.bch.Events().GetLastBlock(ctx)
```
Comment (sshaman1101) [inline insonmnia/oracle/oracle.go]: wrap context with timeout.

**Comment by 3Hren** [comment]
Is this still actual? Seems abandoned.

**Comment by sokel** [comment]
@3Hren yes, is still actual.

File: insonmnia/oracle/oracle.go [hunk 2]
```diff
@@ -85,8 +90,27 @@ func (o *Oracle) listenEventsRoutine(ctx context.Context) error {
...
 	}
-	events, err := o.bch.Events().GetEvents(ctx, o.bch.Events().GetMultiSigFilter(
-		[]common.Address{o.bch.ContractRegistry().OracleMultiSig()}, big.NewInt(0)))
+
+	var lastBlock uint64 = 0
+	var err error
+
+	if o.cfg.Oracle.FromNow {
+		lastBlock, err = func() (uint64, error) {
+			reqCtx, cancel := context.WithTimeout(ctx, 30*time.Second)
```
Comment (sshaman1101) [inline insonmnia/oracle/oracle.go]: consider moving timeout to config or constants

File: insonmnia/oracle/oracle.go [hunk 3]
```diff
@@ -118,36 +144,35 @@ func (o *Oracle) Serve(ctx context.Context) error {
 		zap.String("account", crypto.PubkeyToAddress(o.key.PublicKey).String()),
 		zap.String("price update period:", o.cfg.Oracle.PriceUpdatePeriod.String()),
 		zap.String("contract update period", o.cfg.Oracle.ContractUpdatePeriod.String()),
-		zap.Float64("deviation percent", o.cfg.Oracle.Percent))
+		zap.Float64("deviation percent", o.cfg.Oracle.Percent),
+		zap.Bool("from now", o.cfg.Oracle.FromNow))
```
Comment (sshaman1101) [inline insonmnia/oracle/oracle.go]: zap fields are represented as JSON, so please use underscore char for word splitting.

File: insonmnia/oracle/oracle.go [hunk 4]
```diff
@@ -101,12 +125,14 @@ func (o *Oracle) listenEventsRoutine(ctx context.Context) error {
...
 			case *blockchain.SubmissionData:
-				o.logger.Debug("new submission found", zap.Any("event", event))
-				if o.transactionValid(ctx, value.TransactionId) {
-					o.confirmChanging(ctx, value.TransactionId)
-				} else {
-					o.logger.Info("failed to confirm transaction, transaction invalid ")
+				o.logger.Debug("new submission found", zap.Any("event", *event))
+
+				if !o.transactionValid(ctx, value.TransactionId) {
+					o.logger.Info("failed to confirm transaction, transaction invalid",
```
Comment (sshaman1101) [inline insonmnia/oracle/oracle.go]: Seems like you're stepping into some unusual condition, maybe change log level to a warning?

File: etc/oracle.yaml [hunk 5]
```diff
@@ -15,11 +15,16 @@ blockchain:
...
   #sidechain_endpoint: https://sidechain-dev.sonm.com
   #contract_registry: 0xaf1ffd7f652be7e9a0854a42b2d3046f853f80f1
+  blocks_batch_size: 50000
 
 oracle:
   # In master mode this instance submitting price changing
   # In slave mode this instance confirm changing
   is_master: false
+  #  from_now` parameter forces Oracle to process transactions only from the current block, otherwise, it will read whole transaction history from the beginning of times.
+  #  This parameter works only if slave mode is active.
```
Comment (sshaman1101) [inline etc/oracle.yaml]: I don't spot any sanity check for this condition. What will happen if both `is_master` are `from_now` are set to true?

File: cmd/oracle/main.go [hunk 6]
```diff
@@ -25,19 +25,25 @@ func run(app cmd.AppContext) error {
 		return fmt.Errorf("failed to build logger insance: %s", err)
 	}
 
-	ctx := log.WithLogger(context.Background(), logger)
+	// ctx := log.WithLogger(context.Background(), logger)
```
Comment (sshaman1101) [inline cmd/oracle/main.go]: forgotten comment, probably

File: etc/oracle.yaml (hunk 5 above)
Comment (sokel) [reply etc/oracle.yaml]: in this case `from_now` will be useless.

**Pull request merged**